File: 19.删除链表的倒数第N个节点/removeNthFromEnd.go
package problem0019

type ListNode struct {
	Val  int
	Next *ListNode
}

// 一般思维：找到链表长度，找到n-1位置节点，删除，考虑到删除头节点的情况，需要dummyHead
func removeNthFromEnd(head *ListNode, n int) *ListNode {
	dummy := &ListNode{Next: head}

	temp, length := dummy, 0
	for ; temp != nil; length++ {
		temp = temp.Next
	}
	temp = dummy
	for i := 0; i < length-n-1; i++ {
		temp = temp.Next
	}
	temp.Next = temp.Next.Next

	return dummy.Next
}

// 快慢指针：使得快慢指针相差n，当快指针移动到末尾时，慢指针正好处于倒数第n个节点
func removeNthFromEnd1(head *ListNode, n int) *ListNode {
	dummy := &ListNode{Next: head}

	slow, fast := dummy, dummy
	for i := 0; i < n; i++ {
		fast = fast.Next
	}
	for fast.Next != nil {
		slow = slow.Next
		fast = fast.Next
	}
	slow.Next = slow.Next.Next

	return dummy.Next
}

File: 19.删除链表的倒数第N个节点/removeNthFromEnd.js
// 删除链表的倒数第 N 个结点
// 给你一个链表，删除链表的倒数第 n 个结点，并且返回链表的头结点。你能尝试使用一趟扫描实现吗？
// 1 <= 链表长度 <= 30
// 1 <= n <= 链表长度

// 因为会有删除头节点的情况，因此需要一个dummyHead
// 快慢指针：使得快慢指针保持n的距离，当快指针移动到末尾时，慢指针正好处于倒数第n个节点
function removeNthFromEnd(head, n) {
    const dummy = { next: head }
    let slow = dummy
    let fast = dummy

    for (let i = 0; i < n; i++) {
        fast = fast.next
    }
    while (fast.next !== null) {
        slow = slow.next
        fast = fast.next
    }
    slow.next = slow.next.next
    return dummy.next
}